### DISCUSSION TITLE
is there any reason to not use post_update with mutually dependent relationships?  (answer: no)

### DISCUSSION THREAD

**Discussion by reith** [sqlalchemy/sqlalchemy#9066] [Usage Questions] [answered]
I have two entities Mother and Child, which depend on each other. Based on [Special Relationship Persistence Patterns document](https://docs.sqlalchemy.org/en/20/orm/relationship_persistence.html#post-update), I know I can use `post_update` to issue an update before `DELETE` but I'm not sure if that's really necessary, as I break the relationship manually in code and in some situations it ends up to an `UPDATE` before `DELETE` even without `flush`.

Here is the full sample code:

```python
from typing import List, Optional

import sqlalchemy as sa
from sqlalchemy import orm

engine = sa.create_engine("postgresql+psycopg2://", echo=True)

class Base(orm.DeclarativeBase):
    pass

class Mother(Base):
    __tablename__ = "mothers"

    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    favorite_child_id = orm.mapped_column(sa.ForeignKey("children.id"))
    children: orm.Mapped[List["Child"]] = orm.relationship(
        back_populates="mother",
        foreign_keys="[Child.mother_id]",
        order_by="desc(Child.id)"
    )
    favorite_child: orm.Mapped[Optional["Child"]] = orm.relationship(
        foreign_keys=[favorite_child_id]
    )


class Child(Base):
    __tablename__ = "children"
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    mother_id = orm.mapped_column(sa.ForeignKey("mothers.id"))
    mother: orm.Mapped["Mother"] = orm.relationship(back_populates="children", foreign_keys=[mother_id])


def create_mother_and_child():
    m1 = Mother()
    with orm.Session(engine) as sess:
        sess.add(m1)
        sess.commit()
        sess.refresh(m1)
        m1_id = m1.id

    ch1 = Child()
    ch1.mother = m1
    with orm.Session(engine) as sess:
        sess.add(ch1)
        sess.commit()
        sess.refresh(ch1)
        ch1_id = ch1.id

    with orm.Session(engine) as sess:
        sess.get(Mother, m1_id).favorite_child_id = ch1_id
        sess.commit()

    return (m1_id, [ch1_id])


def kill_the_child_1(mother_id, children_ids):
    with orm.Session(engine) as sess:
        ch1 = sess.get(Child, children_ids[0])
        ch1.mother.favorite_child = None
        sess.delete(ch1)
        sess.commit()


def kill_the_child_2(mother_id, children_ids):
    with orm.Session(engine) as sess:
        ch1 = sess.get(Child, children_ids[0])
        _ = ch1.mother.children
        ch1.mother.favorite_child = None
        sess.delete(ch1)
        sess.commit()


def kill_the_child_3(mother_id, children_ids):
    with orm.Session(engine) as sess:
        ch1 = sess.get(Child, children_ids[0])
        ch1.mother.favorite_child = None
        sess.flush()
        sess.delete(ch1)
        sess.commit()


if __name__ == '__main__':
    Base.metadata.create_all(engine)
    mother_id, children_ids = create_mother_and_child()
    kill_the_child_1(mother_id, children_ids)
    kill_the_child_2(mother_id, children_ids)
    kill_the_child_3(mother_id, children_ids)
```

Here `kill_the_child_1` fails:
```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) update or delete on table "children" violates foreign key constraint "mothers_favorite_child_id_fkey" on table "mothers"
DETAIL:  Key (id)=(3) is still referenced from table "mothers".
```
and `kill_the_child_3` succeeds, because that flush causes SQLAlchemy trigger an `UPDATE` before delete. However, I was surprised when I saw even `kill_the_child_2` caused that `UPDATE` being emitted and the function succeeded. So, what's the magic behind of `_ = ch1.mother.children` and order of generated SQLs and isn't it safe to just `UPDATE` before `DELETE` - without `post_update` being set - since we know objects are both updated and deleted in the code? Links to any documents that might help me learn more about order of generated SQL commands are very much appreciated :)

### Environment:

I observed the same behavior on SQLAlchemy 1.14 and 2.0.0rc1 and the backend is PostgreSQL 10.11.

**Comment by zzzeek** [comment, marked as answer]
You need to use post update.  Sqlalchemy does not support skipping this step for updates see #1063 .

**Reply by reith** [reply]
Thanks Michael for replying.

If I understand correctly, the examples in #1063 presents some cases in which `post_update` may not be theoretically necessary and `flush()` should be enough. My case may be a little bit different, `flush()` works for me even without `post_update`. I'm trying to understand why `flush()` itself is necessary, as in an interesting case, `kill_the_child_2`, I can make SQLAlchemy issue `UPDATE` before `DELETE` by reading the other relationship before changing the second relationship.

**Reply by zzzeek** [reply]
both kill_the_child_1 and kill_the_child_2 perform the same steps, just in a different order.   Run the test against sqlite with foreign key constraints disabled, example one runs DELETE then UPDATE, example 2 runs UPDATE then DELETE.      The reason for the change in order has to do with implementation details, in this case being affected whether or not mother.children is loaded or not.

The point is, the order of the UPDATE / DELETE cannot be defined here because post_update isn't set.   so it's arbitrary.

haven't looked at example 3, because we support this case with post_update being set.   We don't support mutually dependent rows without this setting being set, so it should not matter why it seems to work occasionally or not when the configuration we support is not in place.

**Reply by reith** [reply]
> both kill_the_child_1 and kill_the_child_2 perform the same steps, just in a different order. Run the test against sqlite with foreign key constraints disabled, example one runs DELETE then UPDATE, example 2 runs UPDATE then DELETE.

Yes, and my question was: is it possible and detectable that in these cases always perform UPDATE before DELETE? Look that in the code, I updated the parent model before deleting the child, so it was intuitive to think UPDATE would get executed first.

> The point is, the order of the UPDATE / DELETE cannot be defined here because post_update isn't set. so it's arbitrary.

I think I got a false impression about `post_update`. The documentation says (emphasis is by me):

> To enable the usage of a **supplementary UPDATE statement**, we use the [relationship.post_update](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.post_update) option of [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship). This specifies that the linkage between the two rows should be created using an UPDATE statement after both rows have been INSERTED; it also causes the rows to be **de-associated** with each other via UPDATE before a DELETE is emitted

By "supplementary", I think it means `post_update` is to *add* some `UPDATE`s in some situations, including to de-associate the linkage before deletion. However, the fact that I *de-associate* the linkage myself, (`ch1.mother.favorite_child = None` before `sess.delete(ch1)`) seemed to me I wouldn't need that *supplementary step*.

> haven't looked at example 3, because we support this case with post_update being set. We don't support mutually dependent rows without this setting being set, so it should not matter why it seems to work occasionally or not when the configuration we support is not in place.

I didn't use `post_update` because I thought it's unnecessary as I'm breaking the linkage explicitly. And also, this line in `relationship()` doc:

> Currently, it also will issue an UPDATE after the instance was UPDATEd as well, although this technically should be improved.

Which seemed less deterministic and not cost-free to me.

If nothing, there might be room for documentation improvement. My understanding now is that `post_update` is *necessary* for defining **the order** of `UPDATE` and `DELETE` even if the programmer has the explicit code for `UPDATE`.

Thanks for your input so far.

**Reply by zzzeek** [reply]
> > both kill_the_child_1 and kill_the_child_2 perform the same steps, just in a different order. Run the test against sqlite with foreign key constraints disabled, example one runs DELETE then UPDATE, example 2 runs UPDATE then DELETE.
> 
> Yes, and my question was: is it possible and detectable that in these cases always perform UPDATE before DELETE? Look that in the code, I updated the parent model before deleting the child, so it was intuitive to think UPDATE would get executed first.

the order of operations in the unit of work flush() process is not defined by the order in which you assigned attributes on various objects. nothing about that ordering is captured nor should it be, the only ordering that is captured based on manipulations are the order in which objects of a single class are INSERTed based on session.add() calls, assuming there are no dependencies between those rows directly.  

Otherwise, the order in which operations occur is defined by the dependencies that exist between models.  in this case there is a mutual dependency which requires post_update to be set.


> 
> > The point is, the order of the UPDATE / DELETE cannot be defined here because post_update isn't set. so it's arbitrary.
> 
> I think I got a false impression about `post_update`. The documentation says (emphasis is by me):
> 
> > To enable the usage of a **supplementary UPDATE statement**, we use the [relationship.post_update](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.post_update) option of [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship). This specifies that the linkage between the two rows should be created using an UPDATE statement after both rows have been INSERTED; it also causes the rows to be **de-associated** with each other via UPDATE before a DELETE is emitted
> 
> By "supplementary", I think it means `post_update` is to _add_ some `UPDATE`s in some situations, including to de-associate the linkage before deletion. However, the fact that I _de-associate_ the linkage myself, (`ch1.mother.favorite_child = None` before `sess.delete(ch1)`) seemed to me I wouldn't need that _supplementary step_.

the word "before" in the phrase "i de-associate the linkage myself before" is not captured. If you'd like this ordering to be significant in a manual sense without using post_update, use flush() directly, as you are doing in your kill_the_child_3 example.


> 
> > haven't looked at example 3, because we support this case with post_update being set. We don't support mutually dependent rows without this setting being set, so it should not matter why it seems to work occasionally or not when the configuration we support is not in place.

I've looked at example 3 now and certainly, you can leave post_update out and just flush() when you need to ensure the UPDATE statement has taken place before you attempt a DELETE.  I'm not sure why you wouldn't want to simply use post_update however, which doesn't prevent you from doing the operation manually with a flush() beforehand.


> I didn't use `post_update` because I thought it's unnecessary as I'm breaking the linkage explicitly. 

as long as you call flush() directly, then no it's not needed, however I dont see why one would not want it in place anyway as this makes the model clearer just from a source code perspective as well as operationally.

> And also, this line in `relationship()` doc:
> 
> > Currently, it also will issue an UPDATE after the instance was UPDATEd as well, although this technically should be improved.
> 
> Which seemed less deterministic and not cost-free to me.

Throughout this discussion I've had no idea what issue you had with post_update, this would appear to be the **actual problem**, which is very helpful to know.      This line is not clear as to what it refers towards (which is the same thing described in issue #1063) and does not actually affect performance, I will remove it now.    thanks for clarifying.

**Reply by zzzeek** [reply]
docs will rebuild in an hour or so, the line is removed.  just use post_update as documented.

**Reply by reith** [reply]
Thanks Michael for the thorough explanation and documentation change.
